tests/sm_int/sm_int.c

```diff
@@ -739,24 +739,38 @@ sm_exec_begin
 sm_barrier_begin
 */
 
-#define THREAD_TO_BACK_DELAY 1000
+#define THREAD_DELAY 1000
+
+/*forward*/
+typedef struct SM_RESULT_AND_NEXT_STATE_AFTER_API_CALL_TAG
+{
+    SM_RESULT expected_sm_result;
+    SM_STATES sm_state_after_api;
+}SM_RESULT_AND_NEXT_STATE_AFTER_API_CALL;
 
 typedef struct SM_GO_TO_STATE_TAG
 {
     SM_HANDLE sm;
     uint32_t targetState;
-    HANDLE threadTo;
-    HANDLE threadBack;
+    const SM_RESULT_AND_NEXT_STATE_AFTER_API_CALL* expected;
+    HANDLE threadTo; /*this thread switches the state to the state which is intended to have when sm_..._begin API are called. Then the thread might block (because the state switch is waiting on some draining) or might proceed to end*/
+    HANDLE threadBack; /*this thread unblocks threadTo and the main thread. Main thread might become blocked because the API it is calling might be waiting - such is the case when waiting for a drain to happen*/
+
+    HANDLE targetStateAPICalledInNextLine; /*event set when the target state will be switched in the next line of code. That call might or not return. For example when in the case of wanting to reach the state of SM_OPENED_DRAINING_TO_BARRIER. The call doesn't return until all the sm_exec_end have been called*/
+
+    HANDLE targetAPICalledInNextLine; /*event set just before calling the API in a specific state. This is needed because some of the APIs are blocking (such as calling sm_close_begin when a barrier is executing)*/
 }SM_GO_TO_STATE;
 
-static DWORD WINAPI switchesState(
+
+static DWORD WINAPI switchesTo(
     LPVOID lpThreadParameter
 )
 {
     SM_GO_TO_STATE* goToState = (SM_GO_TO_STATE*)lpThreadParameter;
 
-    LogInfo("set state thread: will now switch state to %" PRI_MU_ENUM "", MU_ENUM_VALUE(SM_STATES, (SM_STATES)(SM_CREATED + goToState->targetState)));
+    LogInfo("switchesTo thread: will now switch state to %" PRI_MU_ENUM "", MU_ENUM_VALUE(SM_STATES, (SM_STATES)(SM_CREATED + goToState->targetState)));
 
+    ASSERT_IS_TRUE(SetEvent(goToState->targetStateAPICalledInNextLine));
     switch (goToState->targetState)
     {
         case 0:/*SM_CREATED*/
```

Comment:
>0 [](start = 13, length = 1)

nit: why not actual enum values like below?

Reply:
reason is poor. Our enums start at '1' with the user-specified values (because _INVALID is "0"). So these numbers have been "indexes" corresponding to enum values. But having the enum values spelled out "as they are" is much better. Changed. 

---
In reply to: [432047519](https://github.com/Azure/azure-c-util/pull/45#discussion_r432047519) [](ancestors = 432047519)